apriltag_ros/src/continuous_detector.cpp

```diff
@@ -86,6 +93,72 @@ bool ContinuousDetector::refreshParamsCallback(std_srvs::Empty::Request& req,
   return true;
 }
 
+void ContinuousDetector::publishEmptyDetection(const std_msgs::Header& header)
+{
+  AprilTagDetectionArray tag_detection_array;
+  tag_detection_array.header = header;
+  tag_detections_publisher_.publish(tag_detection_array);
+}
+
+bool ContinuousDetector::isTagInFOV(const tf::Transform& tag_pose) const
+{
+  if (!camera_model_.initialized())
+  {
+    ROS_WARN_STREAM_THROTTLE(1.0, "Camera not initialized, skipping FOV check");
+    return false;
+  }
+  if (tag_pose.getOrigin().getZ() <= 0)
+  {
+    // handles the case where Z is equal to zero (shouldn't happen) or less than zero (behind the camera)
+    return false;
+  }
+
+  // convert the 3d pose in the image frame to a 2d pixel frame
+  const auto image_pt = camera_model_.project3dToPixel(
+    cv::Point3f(
+      tag_pose.getOrigin().getX(), 
+      tag_pose.getOrigin().getY(), 
+      tag_pose.getOrigin().getZ()
+    )
+  );
+  ROS_DEBUG_STREAM_THROTTLE(1.0, "Frame " << camera_model_.cameraInfo().header.frame_id << " has target at pixel coordinate x: " << image_pt.x << ", y: " << image_pt.y);
+
+  // check if the image is within the target fov (from image width and height in pixels), return true if it is and false otherwise
+  return fov_pixel_buffer_width_ <= image_pt.x && image_pt.x < (camera_model_.cameraInfo().width - fov_pixel_buffer_width_) && fov_pixel_buffer_height_ <= image_pt.y && image_pt.y < (camera_model_.cameraInfo().height - fov_pixel_buffer_height_);
+}
+
+bool ContinuousDetector::isTagInDetectionRange(const tf::Transform& tag_pose) const
+{
+  // distance is the z component in the optical frame
+  const auto curr_dist = tag_pose.getOrigin().getZ();
+  ROS_DEBUG_STREAM_THROTTLE(1.0, "Frame " << camera_model_.cameraInfo().header.frame_id << " has dist to targ: " << curr_dist << ", however min is " << min_detection_dist_ << " and max dist is " << max_detection_dist_);
+  return min_detection_dist_ <= curr_dist && curr_dist <= max_detection_dist_;
+}
+
+void ContinuousDetector::targetPoseCallback(const geometry_msgs::PoseStamped& pose_msg)
+{
+  tf::poseMsgToTF(pose_msg.pose, tag_pose_);
+  // the target pose should be from the base frame
+  base_frame_id_ = pose_msg.header.frame_id;
+}
+
+bool ContinuousDetector::toOptical(const std::string& base_frame, const std::string& optical_frame, tf::Transform& tag_pose)
+{
+  tf::StampedTransform base_to_optical_tf;
+  try
+  {
+    tf_listener_.lookupTransform(base_frame, optical_frame, ros::Time(0), base_to_optical_tf);
```

Comment:
Oh ok. I assumed this was a static transform (like on rzr and rocksteady).... completely forgot about the lifter. In theory, this transform should actually match up with the tag pose transform (since I am multiplying the transform from the tag pose which is provided by the ekf). I'll do some testing but I'll try to take the header from the tag pose message. I was doing that before but in the console I saw the time being 0... so maybe it's not getting set properly on that end. Worst case, I'll change it as you are suggesting